src/lib/device_tree/tests/integration_test.rs

```diff
@@ -71,3 +76,76 @@ fn test_reads_properties() {
     assert_property(it.next().unwrap(), "#address-cells", 2, &vec![0, 0, 0, 1]);
     assert!(it.next().is_none());
 }
+
+#[test]
+fn test_reads_empty_properties() {
+    let data = dts_to_dtb(
+        r#"
+/dts-v1/;
+/ { 
+    #address-cells;
+};"#,
+    );
+    let slice_reader = &SliceReader::new(&data);
+    let reader = FdtReader::new(slice_reader).unwrap();
+    let mut it = reader.walk();
+
+    assert_node(it.next().unwrap(), "", 1);
+    assert_property(it.next().unwrap(), "#address-cells", 2, &vec![]);
+    assert!(it.next().is_none());
+}
+
+#[test]
+fn test_reads_nested_nodes() {
+    let data = dts_to_dtb(
+        r#"
+/dts-v1/;
+/ {
+    node1 {
+        #address-cells = "ok";
+        node2 {
+        };
+    };
+    node3 {
+    };
+};"#,
+    );
+    let slice_reader = &SliceReader::new(&data);
+    let reader = FdtReader::new(slice_reader).unwrap();
+    let mut it = reader.walk();
+
+    assert_node(it.next().unwrap(), "", 1);
+    assert_node(it.next().unwrap(), "node1", 2);
+    assert_property(it.next().unwrap(), "#address-cells", 3, &vec![111, 107, 0]);
+    assert_node(it.next().unwrap(), "node2", 3);
+    assert_node(it.next().unwrap(), "node3", 2);
+    assert!(it.next().is_none());
+}
+
+#[test]
+fn test_returns_error_when_magic_is_invalid() {
+    let mut data = dts_to_dtb(
+        r#"
+/dts-v1/;
+/ {
+    #address-cells = "ok";
+};"#,
+    );
+    // Magic is a first number
+    data[0] = 1;
+    let slice_reader = &SliceReader::new(&data);
+    assert_eq!(FdtReader::new(slice_reader).err(), Some("invalid magic in device tree header"));
```

Comment:
Nit: More direct and idiomatic would be `assert_eq!(FdtReader::new(slice_reader), Err("invalid magic in device tree header"));`. It is a confusing digression to convert from `Result` to `Option` with `.err()`.

Reply:
I believe that would require FdtReader to derive .Debug, PartialEq traits so that assert_eq can print the value in case it is not an error, as well as compare two FdtReaders. I saw using "err()" and comparing with Option as one workaround for it.

Reply:
Oh, I see. Yes. The full type is `Result<FdtReader, &str>` and calling `.err()` allows you to discard the first type so you can compare based on the `&str` alone. Thanks for clarifying.